artiq/frontend/artiq_session.py

```diff
@@ -35,22 +35,25 @@ def main():
     dashboard_cmd += args.d
     ctlmgr_cmd    += args.c
 
-    with subprocess.Popen(master_cmd,
-                          stdout=subprocess.PIPE, universal_newlines=True,
-                          bufsize=1) as master:
-        master_ready = False
-        for line in iter(master.stdout.readline, ""):
-            sys.stdout.write(line)
-            if line.rstrip() == "ARTIQ master is now ready.":
-                master_ready = True
-                break
-        if master_ready:
-            with subprocess.Popen(dashboard_cmd):
-                with subprocess.Popen(ctlmgr_cmd):
-                    for line in iter(master.stdout.readline, ""):
-                        sys.stdout.write(line)
-        else:
-            print("session: master failed to start, exiting.")
+    try:
+        with subprocess.Popen(master_cmd,
+                              stdout=subprocess.PIPE, universal_newlines=True,
+                              bufsize=1) as master:
+            master_ready = False
+            for line in iter(master.stdout.readline, ""):
+                sys.stdout.write(line)
+                if line.rstrip() == "ARTIQ master is now ready.":
+                    master_ready = True
+                    break
+            if master_ready:
+                with subprocess.Popen(dashboard_cmd):
+                    with subprocess.Popen(ctlmgr_cmd):
+                        for line in iter(master.stdout.readline, ""):
+                            sys.stdout.write(line)
+            else:
+                print("session: master failed to start, exiting.")
+    except KeyboardInterrupt:
```

Comment:
Does this close subprocesses cleanly?

Reply:
according to [the docs](https://docs.python.org/3.7/library/subprocess.html#subprocess.Popen) file descriptors will be closed, but no guarantees are given about the subprocess, see also [this discussion](https://bugs.python.org/issue25942). I reverted this change.